File: python/1352. Product of the Last K Numbers.py
class ProductOfNumbers:

    def __init__(self):
        self.total = 0
        self.last_zero = 0
        self.current = []

    def add(self, num: int) -> None:
        self.total += 1
        if not num:
            self.last_zero = self.total
            self.current = []
        else:
            if not self.current:
                self.current.append(num)
            else:
                self.current.append(self.current[-1]*num)

    def getProduct(self, k: int) -> int:
        if self.total - self.last_zero < k:
            return 0
        return (
            self.current[-1]
            if k == len(self.current)
            else self.current[-1] // self.current[-1-k]
        )



# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)

File: java/1352. Product of the Last K Numbers.java
class ProductOfNumbers {

    int total;
    int lastZero;
    List<Integer> current;

    public ProductOfNumbers() {
        this.total = 0;
        this.lastZero = 0;
        this.current = new ArrayList<>();
    }
    
    public void add(int num) {
        total++;
        if (num == 0) {
           lastZero = total;
           current.clear();
        } else {
            if (current.isEmpty()) current.add(num);
            else current.add(current.get(current.size()-1) * num);
        }
    }
    
    public int getProduct(int k) {
        if (total - lastZero < k) return 0;
        if (k == current.size()) return current.get(current.size()-1);
        return current.get(current.size()-1) / current.get(current.size()-1-k);
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.add(num);
 * int param_2 = obj.getProduct(k);
 */